File: frontend/src/types/client-portal.ts
import { CaseDocumentChecklist, CaseQuestionnaire, DocumentRecord, QuestionnaireQuestion } from "@/types/workspace";

export interface ClientPortalAccess {
  case_id: string;
  token_last4: string;
  instructions: string | null;
  expires_at: string | null;
  is_active: boolean;
  last_accessed_at: string | null;
  failed_access_attempt_count: number;
  last_failed_access_at: string | null;
  locked_until: string | null;
  access_session_expires_at: string | null;
}

export interface ClientPortalIssuedAccess extends ClientPortalAccess {
  token: string;
  passcode: string;
  portal_path: string;
}

export interface ClientPortalProgress {
  questionnaire_total_questions: number;
  questionnaire_answered_questions: number;
  questionnaire_percent_complete: number;
  checklist_applicable_items: number;
  checklist_received_items: number;
  checklist_validated_items: number;
  checklist_percent_complete: number;
  overall_percent_complete: number;
}

export interface ClientPortalContext {
  case_id: string;
  case_number: string;
  case_type: string;
  case_title: string;
  case_summary: string | null;
  instructions: string | null;
  access_expires_at: string | null;
  session_token: string;
  session_expires_at: string | null;
  questionnaire: CaseQuestionnaire | null;
  checklist: CaseDocumentChecklist;
  documents: DocumentRecord[];
  progress: ClientPortalProgress;
}

export interface ClientPortalSession {
  session_token: string;
  session_expires_at: string | null;
}

export interface IssueClientPortalAccessInput {
  instructions?: string | null;
  expires_in_days?: number;
}

export interface ClientPortalQuestionnaireValue {
  answer_text?: string | null;
  answer_date?: string | null;
  answer_boolean?: boolean | null;
  answer_choice?: string | null;
  answer_choices?: string[] | null;
  answer_json?: Record<string, unknown> | unknown[] | null;
}

export interface SaveClientPortalAnswerInput {
  token?: string;
  passcode?: string;
  portalSessionToken?: string;
  question: QuestionnaireQuestion;
  value: ClientPortalQuestionnaireValue;
}

export interface UploadClientPortalDocumentInput {
  token?: string;
  passcode?: string;
  portalSessionToken?: string;
  file: File;
  checklistItemId?: string | null;
  documentType?: string | null;
}

File: app/schemas/client_portal.py
import uuid
from datetime import datetime
from typing import Any

from pydantic import BaseModel, Field

from app.schemas.case_questionnaire import (
    CaseQuestionnaireAnswerUpdateRequest,
    CaseQuestionnaireRead,
    QuestionnaireAnswerRead,
)
from app.schemas.document import DocumentRead
from app.schemas.document_checklist import CaseDocumentChecklistRead


class ClientPortalIssueRequest(BaseModel):
    instructions: str | None = None
    expires_in_days: int = Field(default=7, ge=1, le=30)


class ClientPortalAccessRead(BaseModel):
    case_id: uuid.UUID
    token_last4: str
    instructions: str | None
    expires_at: datetime | None
    is_active: bool
    last_accessed_at: datetime | None
    failed_access_attempt_count: int
    last_failed_access_at: datetime | None
    locked_until: datetime | None
    access_session_expires_at: datetime | None


class ClientPortalIssuedRead(ClientPortalAccessRead):
    token: str
    passcode: str
    portal_path: str


class ClientPortalSessionRead(BaseModel):
    session_token: str
    session_expires_at: datetime | None


class ClientPortalAuthRequest(BaseModel):
    token: str = Field(min_length=16)
    passcode: str = Field(min_length=4, max_length=32)


class ClientPortalDocumentUploadResponse(BaseModel):
    document: DocumentRead
    checklist: CaseDocumentChecklistRead


class ClientPortalQuestionAnswerRequest(BaseModel):
    token: str = Field(min_length=16)
    passcode: str = Field(min_length=4, max_length=32)
    payload: CaseQuestionnaireAnswerUpdateRequest


class ClientPortalAnswerCreateRequest(BaseModel):
    token: str = Field(min_length=16)
    passcode: str = Field(min_length=4, max_length=32)
    question_id: uuid.UUID
    value: dict[str, Any]


class ClientPortalProgressRead(BaseModel):
    questionnaire_total_questions: int
    questionnaire_answered_questions: int
    questionnaire_percent_complete: int
    checklist_applicable_items: int
    checklist_received_items: int
    checklist_validated_items: int
    checklist_percent_complete: int
    overall_percent_complete: int


class ClientPortalContextRead(BaseModel):
    case_id: uuid.UUID
    case_number: str
    case_type: str
    case_title: str
    case_summary: str | None
    instructions: str | None
    access_expires_at: datetime | None
    session_token: str
    session_expires_at: datetime | None
    questionnaire: CaseQuestionnaireRead | None
    checklist: CaseDocumentChecklistRead
    documents: list[DocumentRead]
    progress: ClientPortalProgressRead
